src/main/java/frc/robot/subsystems/ColorSensor.java

```diff
@@ -23,24 +30,77 @@
     private final ColorSensorV3 m_colorSensor = new ColorSensorV3(i2cPort);
     private final ColorMatch m_colorMatcher = new ColorMatch();
 
+    private final Color NullTarget = ColorMatch.makeColor(0.0, 0.0, 0.0);
     private final Color BlueTarget = ColorMatch.makeColor(0.143, 0.427, 0.429);
     private final Color GreenTarget = ColorMatch.makeColor(0.197, 0.561, 0.240);
     private final Color RedTarget = ColorMatch.makeColor(0.561, 0.232, 0.114);
     private final Color YellowTarget = ColorMatch.makeColor(0.361, 0.524, 0.113);
 
-    String colorString;
-    
+    private String colorString;
+    private SendableChooser<Color> colorChooser = new SendableChooser<>();
+    private ShuffleboardTab dashboardTab = Shuffleboard.getTab("SmartDashboard");
+    NetworkTableEntry forceFms = Shuffleboard.getTab("SmartDashboard").add("Force FMS", false).withWidget(BuiltInWidgets.kBooleanBox).getEntry();
+ 
+    private Color getColorMatch(){ // actually get real color from fms
+      String gameData;
+      gameData = DriverStation.getInstance().getGameSpecificMessage();
+      if(gameData.length() > 0)
+      {
+        switch (gameData.charAt(0))
+        {
+          case 'B' :
+            return BlueTarget;
+          case 'G' :
+            return GreenTarget;
+          case 'R' :
+            return RedTarget;
+          case 'Y' :
+            return YellowTarget;
+          default :
+            DriverStation.reportWarning("Corrupt Color Data Recieved from FMS", false); //This is corrupt data
+            break;
+        }
+      } else {
+        DriverStation.reportWarning("No Color Data Receieved from FMS", false);
+        //Code for no data received yet
+      }
+      return NullTarget;
+    }
+
+    private void createColorChooser(){
+      SendableRegistry.add(colorChooser, "Color Chooser");
+      SendableRegistry.setName(colorChooser, "Color Chooser");
+
+      colorChooser.setDefaultOption("None", NullTarget);   
+      colorChooser.addOption("Blue", BlueTarget);
+      colorChooser.addOption("Green", GreenTarget);
+      colorChooser.addOption("Red", RedTarget);
+      colorChooser.addOption("Yellow", YellowTarget);
+      
+      dashboardTab.add(colorChooser)
+        .withWidget(BuiltInWidgets.kComboBoxChooser)
+        .withSize(1,1);
+  }
+
     public ColorSensor() {
         m_colorMatcher.addColorMatch(BlueTarget);
         m_colorMatcher.addColorMatch(GreenTarget);
         m_colorMatcher.addColorMatch(RedTarget);
         m_colorMatcher.addColorMatch(YellowTarget);
+        createColorChooser();
     }
 
     @Override
     public void periodic() {
         Color detectedColor = m_colorSensor.getColor();
-
+        Color givenColor;
+        if(DriverStation.getInstance().isFMSAttached() || forceFms.getBoolean(false)){
+          givenColor = getColorMatch(); // this really shouldn't be called periodically 
```

Comment:
Is this from the rules?  Or do we get the color immediately?

Reply:
We only get the color after stage 3 is reached. That's how I interpreted the documentation at least